octree/basic_utils.go

```diff
@@ -92,3 +92,34 @@ func (octree *basicOctree) checkPointPlacement(p r3.Vector) bool {
 		(math.Abs(octree.center.Y-p.Y) <= octree.sideLength/2.) &&
 		(math.Abs(octree.center.Z-p.Z) <= octree.sideLength/2.))
 }
+
+// helperIterateRecursive is a helper function for iterating through a basic octree. If an internal
+// node is found it will be called recursively on each child, updating the idx to correspond to the
+// number of data points in each child. By doing so, each leaf node containing data will have a unique
+// idx value. If the currentBatch number matches the idx%numBatches then the specified function will
+// be performed on the point. If the function returns false, the iteration will end.
+func (octree *basicOctree) helperIterateRecursive(numBatches, currentBatch int, idx uint,
```

Comment:
Visiting the whole tree somewhat defeats the purpose of dividing into batches, since each batch still needs to perform O(n) work. It's true that if the work done at each point is very expensive, this is still useful, but otherwise it will create more work over all. We should instead split the points into contiguous segments and decide which children to visit based on the sizes of the children. Let me know if you want to discuss this.

Reply:
+1